accord-core/src/main/java/accord/coordinate/Exhausted.java

```diff
@@ -37,4 +39,16 @@ public Exhausted(TxnId txnId, @Nullable RoutingKey homeKey, String message)
     {
         super(txnId, homeKey, message);
     }
+
+    protected Exhausted(TxnId txnId, @Nullable RoutingKey homeKey, Exhausted cause)
+    {
+        super(txnId, homeKey, cause);
+    }
+
+    @Override
+    public Exhausted wrap()
```

Comment:
not a fan of this pattern as its hard to maintain... you *must* extend, but the base type also defines, so adding a new type will fail eventually when `wrap` is called... I don't personally see the benefit of keeping the type as it just makes things more confusing to me... `ExecutionException` is a good type as its clear the `cause` is what you want, but also keeps the stack trace of the call... but when we have the main error as the top level type, it can be confusing what happened

Reply:
I can have this generate an illegal state exception if you try to wrap and it changes the type. Alternatively I can restructure hierarchy so there is no default implementation and then it's statically enforced.

The problem with changing the type is that then you have catch blocks that don't work because they don't go through the entire causal chain. That is what originally sent me down this rabbit hole.

Reply:
if you require each type to override `wrap` then `CoordinationFailed` should be `abstract` and this method *must* be `abstract`.  All these checks make maintenance harder

Reply:
I just checked, no one returns `CoordinationFailed`, they all use the higher up types... so making this class abstract should be fine